### ISSUE TITLE
Action button doesn't pop back up after a click

### ISSUE THREAD

**Issue by hadley** [rstudio/shiny#2500]
After running the example in `shiny::actionButton` and clicking on the action button, I see:

<img width="944" alt="image" src="https://user-images.githubusercontent.com/4196/59553302-b189c180-8f58-11e9-8ff6-03ed6ea4e024.png">

This looks like the button is still pressed down; it should pop back up on mouse up.

**Comment by bklingen** [comment]
I agree. Currently, the button reverts to the original state only if one clicks somewhere outside the button. I often want users to repeatedly press the button to trigger some action each time the button is pressed. For that, the current behavior is not very intuitive. (The best I can do is to include a "refresh" icon to signal users  the button can be pressed again.) Ideally, the button would flash (change colors, change border color, etc) for a brief moment when pressed, but then revertsback to the original state.

**Comment by nteetor** [comment]
Do you see this behaviour across browsers? I notice Chrome and the RStudio viewer handle the button focus state differently than FireFox, which has no such state as of now. Perhaps the trick is to hide the added focus state for Chrome and Chrome-like browsers.

**Comment by bklingen** [comment]
Yes, I see this behavior in RStudio viewer, Chrome, Firefox and IE. The button only reverts back to its initial configuration (and changing color on hovering) when first clicking somewhere outside it.

(The behavior also occurs with the nice `actionBttn` provided by the `shinyWidgets` library. Once pressed,  it doens't revert back to the original appearance/color unless one clicks somewhere outside the button.)

**Comment by nteetor** [comment]
The following snippet of JavaScript is a short-term solution,

```javascript 
$(document).on("click", ".action-button", function(e) {
  e.currentTarget.blur();
});
```

If you were willing, I am curious if the `yonder::buttonInput()` has the same problem as the action button or "bttn".

**Comment by trestletech** [comment]
I'm testing with:

```
library(shiny)

ui <- fluidPage(
  actionButton("go", "Go"),
  numericInput("n", "n", 50),
  plotOutput("plot")
)

server <- function(input, output) {

  randomVals <- eventReactive(input$go, {
    runif(input$n)
  })

  output$plot <- renderPlot({
    hist(randomVals())
  })
}

shinyApp(ui, server)
```

I can reproduce the bad behavior in:

 - Chrome (75.0.3770.100)
 - RStudio IDE Viewer (1.2.1226)

I do not see this behavior on:

 - Firefox (67.0.2)
 - Safari (12.1.1)

**Comment by trestletech** [comment]
It looks like the browsers are behaving this way intentionally for accessibility reasons. Relevant conversation: https://github.com/zurb/foundation-sites/issues/6105

The proposal seemed to distill to: 

> whenever a button is focused, the background does not change, regardless of it was focused from keyboard or mouse? So that would mean the only visible indicator you would have that a button is focused, is the blue outline if you used the keyboard to select it?

i.e. preserve the focus ring when navigating by keyboard, but omit it when clicked by mouse. And don't style the button's background color differently on focus. But we will need to force an `outline` since not all browsers offer it by default.

**Comment by trestletech** [comment]
### Option 1

The referenced `what-input` package is an interesting option. Popular npm module with no dependencies, MIT, 4.5kB minified. Enables you to define CSS rules [like](https://github.com/ten1seven/what-input#basic-styling):

```
/*
 * only suppress the focus ring once what-input has successfully started
 */

/* suppress focus ring on form controls for mouse users */
[data-whatintent='mouse'] *:focus {
  outline: none;
}
```

but still will likely require us spelunking through the bootstrap styles.

### Option 2

A lighter-weight solution would just be to blur on-click which, I think, would still preserve accessibility. https://stackoverflow.com/a/25877412/1133019, but might show a flicker of the outline.

### Option 3

There's also a `focus-visible` draft which aims to solve this problem but it's only supported by Firefox, at the moment: https://caniuse.com/#search=focus-visible. There's a [polyfill](https://github.com/WICG/focus-visible) available but it's under a weird license that we'd need to explore more. It's also a dependency-less, small JS file. This route would still require us to specify our own additional CSS rules that leverage this new class to remove the ring. A nice article on the topic: https://fvsch.com/styling-buttons/

### Thoughts

For option 1 or 3, we'd need to consider users who are using custom bootstrap themes. I think we could come up with a universal rule that would undo the outline ring but undoing custom background colors for `focus`ed elements might not be feasible.

**Comment by tsmuse** [comment]
In general hiding the focus outline ring is a dark pattern, the system should always be transparent about where focus is. The issue here seems to be Chromium fires both the click/touch events and the focus events on buttons, unlike Webkit and Gecko. Here are some things to keep in mind for any fixes: 

The ideal way to fix it would be to restyle the focus ring so it doesn't get mistaken for the active state of the button. This leaves the tab-focus systems for each browser intact and consistent (at least across sites that aren't trying to hack the focus ring). While this could be overridden with custom CSS, that's true for pretty much anything we do.

Unfocusing the button on mouseup for buttons that are not toggles will make the UI visually consistent across the three major rendering engines, but it will make Chrome's non-standard tab-focus flow less usable and accessible. The way focus flow works in Chrome is that clicking focuses the element and then tabbing from there picks up from where the user clicked. (In the other browsers the tab-focus is independent of the mouse clicks, so clicking doesn't interrupt where the focus currently is). Blurring the button on mouseup doesn't effect where the user is in the tab order in any browser, so Chrome users will still be changing their place in the tab-order but now they won't have an indication of where they are until they hit tab next.

**Comment by BajczA475** [comment]
I'm only just now finding this thread, but I think I've run into this problem as well, and my (very dumb) workaround was to use shinyjs to disable and re-enable the button back to back upon press inside the button's observer, which seemed to reset its visual aesthetics for me. I'm not sure if this suggests a possible solution, but I thought I'd mention it!

**Comment by gadenbuie** [comment]
At its core, this issue is about the style of the button when the button is focused in the browser. This is handled by Bootstrap in different ways across versions:

| Bootstrap | Normal | Focused |
|:--:|:--:|:--:|
| 3 | ![image](https://github.com/rstudio/shiny/assets/5420529/121562ef-3c94-4654-a82b-8495f787f1ac) | ![image](https://github.com/rstudio/shiny/assets/5420529/32e3552a-b84d-4f77-bc2e-33a641c179a3) |
| 4 | ![image](https://github.com/rstudio/shiny/assets/5420529/13dd684d-1622-4d82-85e2-9c0a23339ffb) | ![image](https://github.com/rstudio/shiny/assets/5420529/24469d85-a581-4173-b1ab-e95bbbe13ccc) |
| 5 | ![image](https://github.com/rstudio/shiny/assets/5420529/13eb66ff-daae-4ddc-847d-00c0d7803aa8) | ![image](https://github.com/rstudio/shiny/assets/5420529/afbb30d9-98b2-43e3-bf22-e7fda876f87f) |

In general, it's least obvious what's going on with Bootstrap 3 and we'd recommend using Bootstrap 4 or 5 if possible by adding `theme = bslib::bs_theme(version = 5)` to your page functions or using [a `page_*()` function from bslib directly](https://rstudio.github.io/bslib/reference/page.html).

I **do not recommend** trying to change the `:focus` or `:active` styles, or using JavaScript or shinyjs to reset the button state or otherwise move the focus away from the button. Doing so reduces your users' ability to use your app with the keyboard or assistive technologies. It's understandable that you'd want to improve your app's visual appearance, but Bootstrap, especially in the most recent versions, provides consistent and careful focus indicators that you can rely on.
